lib/logstash/inputs/gelf.rb

```diff
@@ -127,6 +116,32 @@ def udp_listener(output_queue)
     end
   end # def udp_listener
 
+  # generate a new LogStash::Event from json input and assign host to source_host event field.
+  # @param json_gelf [String] GELF json data
+  # @param host [String] source host of GELF data
+  # @return [LogStash::Event] new event with parsed json gelf, assigned source host and coerced timestamp
+  def self.new_event(json_gelf, host)
+    event = LogStash::Event.new(LogStash::Json.load(json_gelf))
+
+    event[SOURCE_HOST_FIELD] = host
+
+    if (gelf_timestamp = event[TIMESTAMP_GELF_FIELD]).is_a?(Numeric)
+      event.timestamp = self.coerce_timestamp(gelf_timestamp)
+      event.remove(TIMESTAMP_GELF_FIELD)
+    end
+
+    event
+  end
+
+  # transform a given timestamp value into a proper LogStash::Timestamp, preserving microsecond precision
+  # and work around a JRuby issue with Time.at loosing fractional part with BigDecimal.
+  # @param timestamp [String|Numeric] a String, Numeric (integer, float or bigdecimal) representation
+  # @return [LogStash::Timestamp] the proper LogStash::Timestamp representation
+  def self.coerce_timestamp(timestamp)
+    # bug in JRuby prevents correcly parsing a BigDecimal fractional part, see https://github.com/elastic/logstash/issues/4565
+    LogStash::Timestamp.at(timestamp.to_i, BigDecimal.new(timestamp.to_s).frac * 1000000)
```

Comment:
Maybe I've forgot something in the meantime, but why not just use:

```
LogStash::Timestamp.at(BigDecimal.new(timestamp.to_s).to_f)
```

instead of

```
LogStash::Timestamp.at(timestamp.to_i, BigDecimal.new(timestamp.to_s).frac * 1000000)
```

from my tests [here](https://github.com/jruby/jruby/issues/3616) seems to fix the issue too and it's shorter an maybe even faster